modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/rel/IndexScanNode.java

```diff
@@ -119,282 +91,122 @@ public IndexScanNode(
-        throw new UnsupportedOperationException();
-    }
-
-    /** {@inheritDoc} */
-    @Override
-    protected Downstream<RowT> requestDownstream(int idx) {
-        throw new UnsupportedOperationException();
-    }
-
-    private void push() throws Exception {
-        if (isClosed()) {
-            return;
-        }
-
-        checkState();
-
-        if (requested > 0 && !inBuff.isEmpty()) {
-            inLoop = true;
-            try {
-                while (requested > 0 && !inBuff.isEmpty()) {
-                    checkState();
-
-                    RowT row = inBuff.poll();
-
-                    if (filters != null && !filters.test(row)) {
-                        continue;
-                    }
+            Publisher<RowT>[] conditionPublishers = new Publisher[rangeConditions.size()];
 
-                    if (rowTransformer != null) {
-                        row = rowTransformer.apply(row);
-                    }
+            Iterator<RangeCondition<RowT>> it = rangeConditions.iterator();
 
-                    requested--;
-                    downstream().push(row);
-                }
-            } finally {
-                inLoop = false;
-            }
-        }
-
-        if (requested > 0) {
-            if (waiting == 0 || activeSubscription == null) {
-                requestNextBatch();
-            }
-        }
-
-        if (requested > 0 && waiting == NOT_WAITING) {
-            if (inBuff.isEmpty()) {
-                requested = 0;
-                downstream().end();
-            } else {
-                context().execute(this::push, this::onError);
-            }
-        }
-    }
-
-    private void requestNextBatch() {
-        if (waiting == NOT_WAITING) {
-            return;
-        }
-
-        if (waiting == 0) {
-            // we must not request rows more than inBufSize
-            waiting = inBufSize - inBuff.size();
-        }
-
-        Subscription subscription = this.activeSubscription;
-        if (subscription != null) {
-            subscription.request(waiting);
-        } else if (!rangeConditionsProcessed) {
-            RangeCondition<RowT> cond = null;
-
-            if (rangeConditionIterator == null || !rangeConditionIterator.hasNext()) {
-                rangeConditionsProcessed = true;
-            } else {
-                cond = rangeConditionIterator.next();
-
-                rangeConditionsProcessed = !rangeConditionIterator.hasNext();
+            int i = 0;
+            while (it.hasNext()) {
+                conditionPublishers[i++] = indexPublisher(parts, it.next());
             }
 
-            indexPublisher(parts, cond).subscribe(new SubscriberImpl());
+            return SubscriptionUtils.concat(conditionPublishers);
         } else {
-            waiting = NOT_WAITING;
+            return indexPublisher(parts, null);
         }
     }
 
     private Publisher<RowT> indexPublisher(int[] parts, @Nullable RangeCondition<RowT> cond) {
```

Comment:
I would suggest to use `TransformingIterator` here. It will make code much cleaner

Reply:
thanks, fixed